lib/gollum-lib/markups.rb

```diff
@@ -6,7 +6,7 @@ class Markup
         Kramdown::Document.new(content, :auto_ids => false, :input => "markdown").to_html
     }
 
-    register(:markdown,  "Markdown", :regexp => /md|mkdn?|mdown|markdown/)
+    register(:markdown,  "Markdown", :regexp => /md|mkdn?|mdown|markdown/, :h1 => /^(#\s+)(\w+.*)/)
```

Comment:
> By offset you mean [start-line number, end-line number]: good idea!

By `character offset`, I meant the total number of characters before that character (from the start of the source text). Line numbers alone are more complicated to use (see your implementation of `section-data`).

If I had a single class named `EditSectionPageHandle` persisted in the user's session, the solution could be as simple as this:

**tokenize page source (private):**
```ruby
def tokenize(page_source)
    regex_sub_header_content = Gollum::Markup.formats[@markup.format][:h1]
    return data if regex_sub_header_content.nil?

    # this one HAS TO RELATE TO THE PAGE SOURCE, NOT THE TOKENIZED VERSION
    current_section_start_offset = 0

    # tokenize
    page_source.lines.map do |line|
        # replace section content for the token and return it:
        # NOTE: returns "line" if regex didn't match so no problem!
        line.gsub(regex_sub_header_content, "#{current_section_start_offset}")
        
        # before the next iteration:
        # NOTE: "line" contains the whitespaces so no problem!
        current_section_start_offset += line.length
    end.join
end
```

**get section (public):**
```ruby
def get_section(page_source, section_number)
    tokenized = tokenize(page_source)
    
    # TODO: render etc.
    # pseudocode:
    real_sections = nokogiri_doc.get_all_sections   # return all H1s by a css selector, for instance
    
    # section_number is 1-based, so:
    @start_offset = extract_token(real_sections[section_number - 1])
    @end_offset = extract_token(real_sections[section_number])
    if @end_offset.nil?
        # section_number is the last section of the page, in which case:
        @end_offset = page_source.length - 1
    else
        # a little adjustment for the below to work correctly
        @end_offset -= 1
    end
    
    # and return:
    page_source[@start_offset..@end_offset]
end
```

**get edited page (public):**
```ruby
def edited_page(page_source, new_section)
    raise RuntimeError, "The #get_section method has to be invoked first to determine section offsets." if @start_offset.nil? || @end_offset.nil?
    page_source[@start_offset..@end_offset] = new_section
end
```

I think this way, it's much clearer and more concise. And a little faster :).